src/coreclr/jit/compiler.hpp

```diff
@@ -1410,10 +1410,20 @@ inline void GenTree::SetOper(genTreeOps oper, ValueNumberUpdate vnUpdate)
     }
 }
 
+inline void CanonicalizeCastNode(GenTreeCast* cast)
+{
+    if (varTypeIsFloating(cast->CastOp()))
+    {
+        cast->gtFlags &= ~GTF_UNSIGNED;
+    }
+}
+
 inline GenTreeCast* Compiler::gtNewCastNode(var_types typ, GenTree* op1, bool fromUnsigned, var_types castType)
 {
-    GenTreeCast* res = new (this, GT_CAST) GenTreeCast(typ, op1, fromUnsigned, castType);
-    return res;
+    GenTreeCast* cast = new (this, GT_CAST) GenTreeCast(typ, op1, fromUnsigned, castType);
```

Comment:
we are calling GenTreeCast constructor:
https://github.com/dotnet/runtime/blob/419765391c14cb12366ee111b16e91a481ef7bc7/src/coreclr/jit/gentree.h#L3531-L3535
and after that we call `CanonicalizeCastNode` to clear the flag in some cases, is it worth it?
One option is to do 
```
inline GenTreeCast* Compiler::gtNewCastNode(var_types typ, GenTree* op1, bool fromUnsigned, var_types castType)
{
  if (varTypeIsFloating(cast->CastOp())) { fromUnsigned = false; }
  GenTreeCast* cast = new (this, GT_CAST) GenTreeCast(typ, op1, fromUnsigned, castType);
```

or add this logic to `GenTreeCast` constructor (note it will affect more cases then).

Often I set conditional breakpoints on flag values so I would prefer not to change it back and forth without a reason.

Have you tried to add `assert(!varTypeIsFloating(cast->CastOp() || !fromUnsigned)` and fix the callers who call it with wrong value of `bool fromUnsigned`?

Reply:
> Have you tried to add `assert(!varTypeIsFloating(cast->CastOp() || !fromUnsigned)` and fix the callers who call it with wrong value of bool `fromUnsigned`?

Yes, in fact that was the original version of the fix, and the only place I had to fix was the importer. I went back and forth on this a few times, and ultimately decided to have the change initially be with this more conservative implementation, because it is "guaranteed" (not really because I did not add it to the constructor, but nothing calls the constructor directly) to always return "correct" result to the caller. I suppose I should revisit it and run the `assert`ty version through the CI. Maybe it turns out more callers are getting it wrong and it would be better to normalize centrally (regardless of that, the flag flipping will be fixed).

Reply:
> I suppose I should revisit it and run the assertty version through the CI. Maybe it turns out more callers are getting it wrong and it would be better to normalize centrally (regardless of that, the flag flipping will be fixed).

I would prefer this if you have time, if not please open a follow-up issue to investigate/clean it and we will merge this PR as is.

Reply:
> I would prefer this

I would as well, so we'll investigate it as part of this PR.